### ISSUE TITLE
Token

### ISSUE THREAD

**Issue by andye2** [jonassjoh/dspW245#2]
Hi,

thanks for your work and description to figure out the token of the device. I own a W115 and to find the token is a little bit different from the 245.
The config file is in the folger /mnt/user.
There is a OwnerToken and a DeviceToken. It works for me with the device Token, but only for one day. After one day the token will change to another one. Do you have the same behavior with the W245? Do you have a solution for that or a reason why the token change?
Thanks in advance
Andy

**Comment by Garfonso** [comment]
You are right. I thought it would happen after a reboot (and it may happen there, too?) but the token even does change if the device is running. Hm.. :-(

**Comment by jonassjoh** [comment]
> The config file is in the folger /mnt/user

I believe the file is present on multiple locations on the W245 (this could be wrong however, and be my doing). Is there nothing present in the `/mydlink/` folder?

> After one day the token will change to another one.

On the W245 the token persisted as long as I did not reset the device. Could this perhaps somehow be connected to using the mydlink app? Maybe starting the app for the first time will cause a new token to be generated.
Usage of the pin or the token seems to be determined by the `init` value (found in the json returned by requests). If the value is `0`, the pin code is used, if it is not, the device uses the DeviceToken. I'm still somewhat unsure as to what determines the value of `init`. It could be that the values somehow becomes `0` and thus a new token is generated. It may, however, be hard to test if this is the case.

> There is a OwnerToken and a DeviceToken.

I'm not sure what the OwnerToken is used for.

**Comment by jonassjoh** [comment]
As for a workaround you could possibly control the device through the terminal instead.
I believe you can toggle the outlet through the use of the command `GpioForCrond <S> <O>` where `<S> - 0=off, 1=on` and `<O> - 11/17/19/18 (outlet 1/2/3/4)` (you may have to use different values for <O> on the W115).

**Comment by Garfonso** [comment]
In my [library](https://github.com/Garfonso/dlinkWebSocketClient) I added a script (and a flag) to get the token from telnet automatically. (if you like to use node.js, you can have a look)

But I think for w115 it is probably best to not pair it with the app and use the PIN (I will try that again, soonish). 

@jonassjoh 
Do you know how to prevent a firmware update (that for example might change credentials for the telner access)?

**Comment by andye2** [comment]
Thanks for the feedback. On the w115 there is no mydlink folder. 
And yes, the device is connected to the mydlink app. And I can imagine that the change will be a command for security reason. Dlink will trigger that. But why is the w245 so different. Curious...

**Comment by Garfonso** [comment]
> 
> 
> Thanks for the feedback. On the w115 there is no mydlink folder.
> And yes, the device is connected to the mydlink app. And I can imagine that the change will be a command for security reason. Dlink will trigger that. But why is the w245 so different. Curious...

Uhh.. that's strange. On my W115 there is a /mydlink folder. It is right in the root of the filesystem (hence the / at the beginning). But you are right, /mnt/user is the same contents as /mydlink/config on my device.

**Comment by jonassjoh** [comment]
> Do you know how to prevent a firmware update (that for example might change credentials for the telner access)?

There should be an option in the mydlink app. I'm not sure how to do it otherwise. However, if you boot the device into recovery mode there should be a webserver on port 80 with a website that allows you to upload your own firmware image. So it should be no issue to revert back to an older firmware. Older firmware images can be found on the D-Link website.

**Comment by jonassjoh** [comment]
Here is a list of possible commands that you can send. I'm not sure if they will all work, but they are present in the code atleast. I also noticed that `set_fw_auto_upgrade` is present in the list, which might be of interest to you considering your other comment.

```
sign_in
av_config
get_setting
set_setting
status
get_status
start_viewing
stop_viewing
register
recycle_channel
get_policy
get_schedule
change_policy
change_schedule
change_scene
event
unbind_device
update_info
device_reset
reload_config
get_server_info
set_service_level
get_service_level
keep_alive
upload_log
voice_report
get_unit_data
get_setup_status
wlan_survey
test_token
get_regular
conn_init
conn_add_candidate
rtsp_command
rtsp_start_viewing
conn_close
conn_list
device_status
update_cnvr_state
sd_list_event
sd_list_snapshot
sd_list_file
fw_upgrade
m2m_event
rtsp_ack_channel
set_fw_auto_upgrade
get_fw_auto_upgrade
change_policy_batch
change_schedule_batch
change_scene_batch
```

**Comment by andye2** [comment]
> > Thanks for the feedback. On the w115 there is no mydlink folder.
> > And yes, the device is connected to the mydlink app. And I can imagine that the change will be a command for security reason. Dlink will trigger that. But why is the w245 so different. Curious...
> 
> Uhh.. that's strange. On my W115 there is a /mydlink folder. It is right in the root of the filesystem (hence the / at the beginning). But you are right, /mnt/user is the same contents as /mydlink/config on my device.

Ok, I have explanation why the folder is not visible. I use the recovery mode to figure out the token. In this mode there is no mydlink folder. 
But if the device is connected to your home wifi and you login via telnet then you have the mydlink folder. but anyway, the content is the same.

And I don´t think the firmware upgade is a problem. I started with Version 1.00-b11, and now it is 1.02-b04, but the behavior is still the same.

**Comment by xses79** [comment]
> > > Thanks for the feedback. On the w115 there is no mydlink folder.
> > > And yes, the device is connected to the mydlink app. And I can imagine that the change will be a command for security reason. Dlink will trigger that. But why is the w245 so different. Curious...
> > 
> > 
> > Uhh.. that's strange. On my W115 there is a /mydlink folder. It is right in the root of the filesystem (hence the / at the beginning). But you are right, /mnt/user is the same contents as /mydlink/config on my device.
> 
> Ok, I have explanation why the folder is not visible. I use the recovery mode to figure out the token. In this mode there is no mydlink folder.
> But if the device is connected to your home wifi and you login via telnet then you have the mydlink folder. but anyway, the content is the same.
> 
> And I don´t think the firmware upgade is a problem. I started with Version 1.00-b11, and now it is 1.02-b04, but the behavior is still the same.

Did you get this to work?

**Comment by Garfonso** [comment]
What is your question?

W115 works. But with the app connected the token changes often, approximately daily.

**Comment by xses79** [comment]
> What is your question?
> 
> W115 works. But with the app connected the token changes often, approximately daily.

Okay, so that will be a problem, i also got no clue where and how to use 

> What is your question?
> 
> W115 works. But with the app connected the token changes often, approximately daily.

So it would be best to remove it from the App?, but i still do not understand where and how i should get this running in  Home Assistant, i can do in terminal "python3 dswpW245" and i get "sending for http upgrde" "sign in for salt" and "Turning led 1 on" but thees are just text read-back? nothing happens with the unit, not even the same day i i got the tokens out by telnet on the unit. 
I am a novice at this only been playing with it for a couple of weeks, but i have 6 W115 and 3 W118 i would like to use.

**Comment by jonassjoh** [comment]
You're not supposed to run the script directly. Doing so will only run some code meant to act as an example. Try creating a new script where you import this one instead. See the readme for example code.

Also, I do not know how Home Assistant works and the code has **not** been written to be compatible with it. Perhaps someone else can be of assistance regarding this.
